pkg/util/objectpool/withoutlock/objectpool.go

```diff
@@ -0,0 +1,210 @@
+func NewPoolWithSpec(spec Spec) *Pool {
+	p := &Pool{
+		initSize:     spec.initSize,
+		maxSize:      spec.maxSize,
+		new:          spec.new,
+		checkWhenPut: spec.checkWhenPut,
+		checkWhenGet: spec.checkWhenGet,
+		store:        make(chan IPoolObject, spec.maxSize),
+	}
+	if err := p.init(); err != nil {
+		logger.Errorf("new pool failed %v", err)
+		return nil
+	}
+	return p
+}
+
+// Validate validate
+func (s *Spec) Validate() error {
+	if s.initSize > s.maxSize {
+		s.maxSize = s.initSize
+	}
+	if s.maxSize <= 0 {
+		return fmt.Errorf("pool max size must be positive")
+	}
+	if s.initSize < 0 {
+		return fmt.Errorf("pool init size must greate than or equals 0")
+	}
+
+	return nil
+}
+
+// initializes the pool with initSize of objects
+func (p *Pool) init() error {
+	for i := 0; i < int(p.initSize); i++ {
+		iPoolObject, err := p.new()
+		if err != nil || !iPoolObject.HealthCheck() {
+			logger.Errorf("create pool object failed or object of pool is not healthy when init pool:, err %v", err)
+			continue
+		}
+		atomic.AddInt32(&p.size, 1)
+		p.store <- iPoolObject
+	}
+	return nil
+}
+
+// Get returns an object from the pool,
+// if no free object, it will create a new one if the pool is not full
+// if the pool is full, it will block until an object is returned to the pool
+func (p *Pool) Get(ctx context.Context) (IPoolObject, error) {
+	backoff := 1
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case poolObject := <-p.store:
+			//fmt.Printf("[%d] id %d size %d got oj from channel\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+			if p.checkWhenGet && !poolObject.HealthCheck() {
+				//fmt.Printf("[%d] id %d size %d would destroy oj\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+				p.destroyIPoolObject(poolObject)
+				if poolObject = p.createIPoolObject(); poolObject != nil {
+					//fmt.Printf("[%d] id %d size %d created new ob after destroy\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+					return poolObject, nil
+				}
+				continue
+			}
+			//fmt.Printf("[%d] id %d size %d checked ob from channel\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+			return poolObject, nil
+		default:
+			//fmt.Printf("[%d] id %d size %d enter default case\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+			if iPoolObject := p.createIPoolObject(); iPoolObject != nil {
+				//fmt.Printf("[%d] id %d size %d created new oj when default case\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+				return iPoolObject, nil
+			}
+			//fmt.Printf("[%d] id %d size %d would go backoff\n", time.Now().UnixMicro(), goID(), atomic.LoadInt32(&p.size))
+			for i := 0; i < backoff; i++ {
+				runtime.Gosched()
+			}
+			if backoff < maxBackoff {
+				backoff <<= 1
+			}
+		}
+	}
+}
+
+func (p *Pool) createIPoolObject() IPoolObject {
+	if atomic.AddInt32(&p.size, 1) > p.maxSize {
+		atomic.AddInt32(&p.size, -1)
+		return nil
+	}
```

Comment:
```suggestion
	for {
		size := p.size
		if size > p.maxSize {
			return nil
		}
		if atomic.CompareAndSwapInt32(&p.size, size, size + 1) {
			break
		}
	}
```

Reply:
I have considered like that, but in some scenarios, for example, the `new` function always return a unhealthy object, then `pool.size` could be a dynamic balance: `pool.size` always less than `pool.maxsize` and `pool.size` is always changing, so some goroutines will fall into this for loop. 

My code can at least guarantee that created ipoolobject num will not  more than pool.maxsize. Of course, its calculations will not be so accurate in turn, and it will miss the opportunity for it to enter the subsequent process, but at least it will have a chance in the next cycle. Overall in a race, I want to return to the case of context.timeout as much as possible, rather than being stuck in a loop.

Reply:
> I have considered like that, but in some scenarios, for example, the new function always return a unhealthy object, then pool.size could be a dynamic balance: pool.size always less than pool.maxsize and pool.size is always changing, so some goroutines will fall into this for loop.

I'm not getting your point clearly, but I think what you worry about won't happen: using CAS in a loop is a common pattern in lock-free programming.